### ISSUE TITLE
NCPA 1.8.1 can't handle multiple checks

### ISSUE THREAD

**Issue by guglez** [NagiosEnterprises/ncpa#239]
Hi,

I'm using ncpa 1.8.1 for Windows. We are checking mssql DB via custom powershell script. At the very beginning we had only 5 checks and everything was fine. Now we have 40+ checks and if I will fire all checks then I will get errors for every check. I tried to run it with -v option and I can see this:

An error occurred:[Errno 104] Connection reset by peer

I tried to increase timeout everywhere (-T option,timeouts at Nagios Core Configuration section) but has no luck. I believe it's a bug. I found similar issue here - https://support.nagios.com/forum/viewtopic.php?f=16&t=36635&start=10 but I also don't know where I can download the old version.

WBR

**Comment by jhomann** [comment]
This is a duplicate of issue #219 and can be resolved by downloading the latest check_ncpa.py and using that. You can download it here: https://raw.githubusercontent.com/NagiosEnterprises/ncpa/master/client/check_ncpa.py

Let me know how this goes.

**Comment by guglez** [comment]
That shoes not help. I have the following message as a result of the ckeck:
UNKNOWN: Execution exceeded timeout threshold of 60s


Which timeout is it (nagios core config or plugin or something else)?

**Comment by jhomann** [comment]
The connection reset by peer issue was caused by the way urllib worked but now by default there is a  -T option in `check_ncpa.py` that is set to 60 seconds and is probably what you are hitting. The new 2.0.0 of NCPA will also have a default 60 second timeout in the NCPA passive agent itself which will be able to be changed in the `ncpa.cfg` file.

**Comment by guglez** [comment]
I forgot to mention that this action is causing 100%cpu load on a host where NCPA is installed. I do not expect this because of the script. It's very simple. It executes simple query on mssql database and checks the result. If I run this script by hand it returns a result instantly. But when I run it via NCPA it takes sometimes 45 seconds and it's also causing CPU load.

**Comment by jhomann** [comment]
What kind of script is it? Is it powershell? There have been instances where powershell doesn't work quite right because of it requiring more commands passed to it than what we pass in the `[plugin directives]` section of `ncpa.cfg`. However I am not that familiar with powershell.

**Comment by ccztux** [comment]
If you use Windows 64 bit edition, then have a look to this:
[Powershell 64 bit](https://github.com/NagiosEnterprises/ncpa/issues/192#issuecomment-219275286)

**Comment by guglez** [comment]
@jhomann I'm also new to powershell. It's a script which launches Invoke-Sqlcmd with some parameters. Can I restrict somehow the amount of checks which can be executed in parallel?

@ccztux I tried both 32 and 64 bit powershells but results are the same.

**Comment by jhomann** [comment]
At the moment you cannot determine how many can be run in parallel although this might be useful for people in the future so I will definitely have to keep it in mind. As for what is keeping the CPU maxed, I am curious if the wait in python causes that ... it would be very strange indeed if it did though. I've never heard of such a thing happening. I can try to test this but it may take some time.

**Comment by guglez** [comment]
@jhomann as far as I can see powershell processes are eating all CPU

**Comment by jhomann** [comment]
So it's not the ncpa .exe that is using the CPU it's the actual powershell.exe running the script?

**Comment by guglez** [comment]
@jomann09 in my case my powershell scripts are not eating cpu at all when I execute them by hand. They eat CPU only when I trigger them via ncpa.

But I need to state that I do not have expert Microsoft administration skills, I'm a linux Engineer.

**Comment by jomann09** [comment]
I'd like to ask you how often these checks are being ran. They are active checks right? I'd also like to see your `check_command` if you can strip out anything that you can't share!

**Comment by guglez** [comment]
We have about 40 checks. I had to run them every 30 minutes. Otherwise there is a chance that they will eat 100% cpu and return timeouts.

T 60 -H IPADDRESS_OF_MY_HOST -t 'MY_TOKEN' -P PORT# -M 'agent/plugin/checkbatch.ps1' -a "-warn '599' -crit '600' -srvinstance 'here some mssql params' -dbname 'here some mssql params' -query_name 'Order received' -dbusername 'user' -dbpw 'password'"

**Comment by walliscz** [comment]
guglez: I had a similar problem, I solved by increasing timeout in nagios.cfg file from 60 to 180s.

vi /usr/local/nagios/etc/nagios.cfg
service_check_timeout=180

**Comment by jomann09** [comment]
Which part did you also have @walliscz ? The original error or that your powershell scripts are taking 100% cpu?

**Comment by walliscz** [comment]
With this:

> That shoes not help. I have the following message as a result of the check:
UNKNOWN: Execution exceeded timeout threshold of 60s
Which timeout is it (nagios core config or plugin or something else)?

**Comment by jomann09** [comment]
If you're using the latest plugin (check_ncpa.py) with NCPA 1.8.1 then it is most likely the plugin that is timing out. You can use `-T <timeout in seconds>` with the latest ([shown here](https://github.com/NagiosEnterprises/ncpa/blob/master/client/check_ncpa.py#L88)) check_ncpa.py plugin

**Comment by jomann09** [comment]
I've actually added `bufsize=-1` to the `subprocess.Popen` call in the plugin node due to the Python docs mentioning that if there are performance issues, to run using a buffer. The `-1` value tells it to use the system's default settings. I've tested it on Windows and Linux and seems to be working fine, so this may help with this issue, although I am still unable to cause it on my own systems.

**Comment by jomann09** [comment]
Closing this for now because I still have yet to be able to recreate anything along the lines of using 100% CPU. Although I am using the lastest code in the dev branch for testing.

I even made a powershell script that just does a loop over a million numbers but that only pegged the CPU at around 10% for a good minute.
